internal/service/resources/collector/collector.go

```diff
@@ -0,0 +1,229 @@
+		case <-time.After(time.Minute * 1):
+			c.execute(ctx)
+		case <-ctx.Done():
+			slog.Info("Context terminated; collector exiting")
+			return nil
+		}
+	}
+}
+
+// init runs the onetime initialization steps for the collector
+func (c *Collector) init(ctx context.Context) error {
+	slog.Info("initializing data collector")
+
+	for _, d := range c.dataSources {
+		err := c.initDataSource(ctx, d)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// initDataSource initializes a single data source from persistent storage.  This recovers its unique UUID and
+// generation ID so that it continues from its last save point.
+func (c *Collector) initDataSource(ctx context.Context, dataSource DataSource) error {
+	name := dataSource.Name()
+	record, err := c.repository.GetDataSourceByName(ctx, name)
+	switch {
+	case errors.Is(err, utils.ErrNotFound):
+		// Doesn't exist so create it now.
+		result, err := c.repository.CreateDataSource(ctx, &models.DataSource{
+			Name:         name,
+			GenerationID: dataSource.GetGenerationID(),
+		})
+		if err != nil {
+			return fmt.Errorf("failed to create new data source %q: %w", name, err)
+		}
+
+		dataSource.SetID(*result.DataSourceID)
+		slog.Info("created new data source", "name", name, "uuid", *result.DataSourceID)
+	case err != nil:
+		return fmt.Errorf("failed to get data source %q: %w", name, err)
+	default:
+		dataSource.SetID(*record.DataSourceID)
+		dataSource.SetGenerationID(record.GenerationID)
+		slog.Info("restored data source",
+			"name", name, "uuid", record.DataSourceID, "generation", record.GenerationID)
+	}
+	return nil
+}
+
+// execute runs a single iteration of the main loop.  It does not return an error because all errors should be handled
+// gracefully.  If a truly unrecoverable error happens then a panic should be used to restart the process.
+func (c *Collector) execute(ctx context.Context) {
+	slog.Debug("collector loop running", "sources", len(c.dataSources))
+	for _, d := range c.dataSources {
+		slog.Debug("collecting data from data source", "source", d)
+		if err := c.executeOneDataSource(ctx, d); err != nil {
+			slog.Warn("failed to collect data from data source", "source", d, "error", err)
+		} else {
+			slog.Debug("collected data from data source", "source", d)
+		}
+	}
+	slog.Debug("collector loop complete", "sources", len(c.dataSources))
+}
+
+// executeOneDataSource runs a single iteration of the main loop for a specific data source instance.
+func (c *Collector) executeOneDataSource(ctx context.Context, dataSource DataSource) (err error) {
+	// Increment the generation id
+	dataSource.IncrGenerationID()
+
+	// Get the list of resource pools for this data source
+	pools, err := c.collectResourcePools(ctx, dataSource)
+	if err != nil {
+		return fmt.Errorf("failed to collect resource pools: %w", err)
+	}
+
+	// Get the list of resources for this data source
+	_, err = c.collectResources(ctx, dataSource, pools)
```

Comment:
dataSource.GetResources doesn't need pools, so we could delete it. If we do that, is it not safe to run collectResources even if collectResourcesPools fails. We could attempt to sync Resources

Reply:
The reason I set this up this way is that it will have to be this way in the future for the h/w manager because as of now there is no way to get just a full list of resources, they need to be scoped by resource pool... so for the ACM one it just gets ignored, but other data source types may need this information.